pytact/data_reader.py

```diff
@@ -568,53 +568,43 @@ def get_index(index):
-
-    def _clustered_global_context(self, seen: set[Node]) -> Iterable[list[Definition]]:
-        if prev := self.previous:
-            cluster = list(prev.cluster)
-            yield cluster
-            yield from cluster[-1]._clustered_global_context(seen)
-        for eprev in self.external_previous:
-            if eprev.node not in seen:
-                cluster = list(eprev.cluster)
-                yield cluster
-                yield from cluster[-1]._clustered_global_context(seen)
-                seen.add(eprev.node)
-
-    @property
-    def global_context(self) -> Iterable[Definition]:
+    def global_context(self, across_files : bool = True, inclusive = False, seen : set[Node] = None) -> Iterable[Definition]:
         """All of the definitions in the global context when this definition was created.
 
         Note that this does not include this definition itself, except when the definition is a inductive,
         constructor or projection. Because those are mutually recursive objects, they reference themselves
         and are therefore part of their own global context.
         """
-        start = self.cluster_representative
-        match start.kind:
-            case Definition.Inductive(_) | Definition.Constructor(_) | Definition.Projection(_):
-                yield start
-        yield from start._global_context(set())
+        d = self.cluster_representative
+        if seen is None: seen = set()
 
-    @property
-    def clustered_global_context(self) -> Iterable[list[Definition]]:
+        # force inclusivity for recursive definitions
+        match d.kind:
+            case Definition.Inductive(_) | Definition.Constructor(_) | Definition.Projection(_):
+                inclusive = True
+        
+        while d is not None:
+            if inclusive: yield d # skip first result if not inclusive
+            else: inclusive = True
+            if across_files:
+                for eprev in d.external_previous:
+                    if eprev.node in seen: continue
+                    yield from eprev.global_context(
+                        inclusive = True,
+                        across_files = True,
+                        seen = seen,
+                    )
+                    seen.add(eprev.node)
+
+    def clustered_global_context(self, across_files : bool = True, inclusive : bool = False) -> Iterable[list[Definition]]:
         """All of the definitions in the global context when this definition was created, clustered into
         mutually recursive cliques.
 
         The definition itself may be part of the first mutually recursive cluster."""
-        start_cluster = list(self.cluster)
-        match start_cluster[0].kind:
-            case Definition.Inductive(_) | Definition.Constructor(_) | Definition.Projection(_):
-                yield list(start_cluster)
-        yield from start_cluster[-1]._clustered_global_context(set())
+        return self._build_clusters(self.global_context(
```

Comment:
Maybe call it `_split_by_clusters`?

Reply:
why not, I didn't know a really good fitting name for it

Reply:
`group_by_clusters` ?

Reply:
Sounds good.